### PR TITLE
LoRA Finetuning

### PR DESCRIPTION
**Pull request by riensou**
#26

### CODE REVIEW

**Comment by riensou** [comment]
examples of some training runs:

1. training a gpt2 model on linear regression from scratch

train_linear


2. lora fine-tuning the above model on sparse linear regression

lora_sparse

3. training a gpt2 model on sparse linear regression from scratch

train_sparse

4. lora fine-tuning the above model on linear regression

lora_linear

File: src/parse/trainer.py [hunk 1]
```diff
@@ -146,9 +149,21 @@ def _produce_trainer_stages(data: dict) -> TrainerSteps:
 
     model = get_model(stages[0]['model'] | { "x_dim" : x_dim }, x_dim, y_dim)
     optimizer = get_optimizer(model, stages[0]['optim'])
-    if 'model_weights' in data and 'optim_state' in data:
+
+    resume = 'model_weights' in data and not 'base_model' in stages[0]['model']
+    lora = 'base_model' in stages[0]['model'] and type(model) == MODELS['lora']
```
Comment (riensou) [inline src/parse/trainer.py:154]: this line worries me, but I need it in order to make sure that `model.setup_peft_model (...)` exists.

File: src/parse/trainer.py [hunk 2]
```diff
@@ -146,9 +149,21 @@ def _produce_trainer_stages(data: dict) -> TrainerSteps:
...
+    lora_resume = lora and 'model_weights' in data and any(['lora' in weight for weight in list(data.get('model_weights', {}).keys())])
```
Comment (riensou) [inline src/parse/trainer.py:155]: this line also worries me, because it has the underlying assumption that the string 'lora' will be in the finetuning weights, which is not a scalable solution for other finetuning methods, but I'm not sure how to solve this problem otherwise

File: src/models/lora.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+                self.load_state_dict(base_model_weights)
+            for param in self.parameters():
+                param.requires_grad = False
+            self._backbone = get_peft_model(self._backbone, self.lora_config) # pyright: ignore[reportArgumentType]
+        elif lora_model_weights:
+            self._backbone = get_peft_model(self._backbone, self.lora_config) # pyright: ignore[reportArgumentType]
+            self.load_state_dict(lora_model_weights)
+            for weight in lora_model_weights:
+                if 'lora' not in weight:
+                    lora_model_weights[weight].requires_grad = False
```
Comment (riensou) [inline src/models/lora.py:35]: I need to verify the correctness of this, will setting requires_grad to False here actually work? I printed out a bunch of parameters.requires_grad values, and most were False except for some Trues, which is promising, but more rigorous testing is needed

File: src/models/lora.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
+from peft import LoraConfig, get_peft_model # pyright: ignore[reportPrivateImportUsage]
+from typing import Optional, Any
+from torch import nn
+
+from .transformer import TransformerModel
+from core import ContextModel
+
+class Lora(TransformerModel):
```
Comment (riensou) [inline src/models/lora.py:8]: should we have a `FineTuner` class that is a child of `TransformerModel`, then rename this file to finetune.py. Then make `Lora` a child of `FineTuner`. That way, we can add other finetuning methods in the future (i.e. linear probing, model-stitching, etc.)

File: src/parse/trainer.py (hunk 1 above)
Comment (nelson-lojo) [reply src/parse/trainer.py:154]: Perhaps a nesting structure would help:
```
{
    "optimizer_state": ...,
    "model_weights": <ContextModel weights>,
    "fine_tune": {
        "model_weights": <Adapter 1 weights>,
        "fine_tune": {
            "model_weights": <Adapter 2 weights>,
        }
    }
}
```

File: src/models/lora.py (hunk 3 above)
Comment (nelson-lojo) [reply src/models/lora.py:35]: iirc this is the in-place way of removing tensors from the computation graph ... it should be identical to [`<Tensor>.detach()`](https://pytorch.org/docs/stable/generated/torch.Tensor.detach.html)

File: src/parse/trainer.py (hunk 2 above)
Comment (riensou) [reply src/parse/trainer.py:155]: not sure what you mean here, as the `model_weights` are gotten from the CL, and the nested structure could only be created by doing the logic I just did, right? unless we change the interaction with the CL or yml files?
